### PR TITLE
feat(phyai): add LingBot V2 inference

### PR DESCRIPTION
**Pull request by An-mo-9**
## Summary

This PR adds end-to-end LingBot-VLA 2.0 inference support to PhyAI.

- Add the LingBot V2 model plugin, including:
  - Qwen3-VL vision encoder with DeepStack features
  - Qwen3-VL language backbone with multimodal MRoPE
  - learned current and future perceptual query prefixes
  - Qwen2 action expert with token-level MoE and a ten-step flow-matching Euler schedule
  - BF16 parameters and residual streams with the released FP32 routing and joint-attention boundaries
  - CUDA Graph execution for the fixed-shape action-expert Euler loop
- Use the released Conv3D PatchEmbed operator by default.
- Provide a mathematically equivalent GEMM PatchEmbed backend as an explicit optimization option.
- Reuse FlashInfer's existing CUTLASS fused MoE backend for routed expert execution.
- Add reusable LingBot V2 preprocessing and postprocessing in `phyai-utils-tools`.
- Add a local three-camera inference example with deterministic inputs and action export.
- Add fixed-contract latency, component latency, and MFU benchmark tools.
- Support BF16 or FP32 vision parameters and Conv3D or GEMM PatchEmbed selection through `LingBotV2Args`.

## Checkpoint loading

Released LingBot V2 checkpoints can be loaded without conversion in either form:

- a single safetensors file
- sharded safetensors with `model.safetensors.index.json`

Checkpoint loading remains strict and reports missing or unexpected weights.

## Validation

The implementation was validated with the released LingBot V2 checkpoint:

- Strict sharded checkpoint loading succeeded.
- End-to-end inference produced finite actions with shape `(1, 50, 55)`.
- The Conv3D and GEMM PatchEmbed paths preserve the same model shapes, checkpoint mapping, and BF16 output dtype.
- NVIDIA Thor inference completed with the explicit GEMM PatchEmbed option.

## Tests

The focused LingBot V2 processor, configuration, model, precision-policy, runtime, benchmark, and CUDA Graph tests passed:

```text
74 passed
```

## Summary by CodeRabbit

* **New Features**
  * Added end-to-end LingBot V2 vision-language-action inference support.
  * Added configurable image, robot-state, normalization, noise, device, and output handling.
  * Added RoboTwin integration for observation/action conversion and WebSocket policy serving.
  * Added tools for paired evaluations, resumable task runs, and CSV/Markdown success reports.
  * Added FP32 routing support for improved numerical stability.

* **Documentation**
  * Added RoboTwin setup, evaluation, and protocol guidance.

* **Tests**
  * Added coverage for preprocessing, normalization, inference behavior, protocol defaults, and numerical accuracy.

### CODE REVIEW

File: benchmark/lingbot_v2/check_cuda_graph_parity.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+    config = load_config(args.checkpoint, LingBotVLA2Config)
+    tensors, input_metadata = load_inputs(args.input)
+    contract = validate_contract(tensors, config)
+    active_patch_counts = tensors["input.image_grid_thw"].prod(dim=-1)[
+        tensors["input.image_masks"].bool()
+    ]
+    max_vision_tokens = int(
+        (active_patch_counts // config.vision.spatial_merge_unit).max()
+    )
```
[the remark this answers is not in the document]
Comment (An-mo-9) [reply benchmark/lingbot_v2/check_cuda_graph_parity.py]: if image_masks.shape != (batch_size, num_images) or not bool(image_masks.all()):
    raise ValueError("all three benchmark camera views must be active.")
   
   已经检查过全为flase的情况

File: benchmark/lingbot_v2/compare_moe_thor.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+            hidden,
+            weights,
+            robby_moe_forward,
+            top_k=6,
+            routed_scaling_factor=1.0,
+            n_warmup=args.n_warmup,
+            n_timed=args.n_timed,
+        )
+        for name, hidden in hidden_inputs.items()
+    }
```
[the remark this answers is not in the document]
Comment (An-mo-9) [reply benchmark/lingbot_v2/compare_moe_thor.py]: 这里确实写错了，lingbotv2中top_k=4,routed_scaling_factor=4.0，优化时忘改回来了，已修正

File: benchmark/lingbot_v2/compare_thor_latency.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+        "mean_ms": latency["mean"],
+        "p50_ms": latency["p50"],
+        "p90_ms": latency["p90"],
+        "p99_ms": latency["p99"],
+        "std_ms": latency["std"],
+        "peak_allocated_gib": (int(checks["allocated_peak_bytes"]) / 2**30),
+        "params_dtype": meta["params_dtype"],
+        "vision_dtype": meta["vision_dtype"],
+        "attention_backend": meta.get("attention_backend", "PHYAI"),
+        "moe_backend": meta.get("moe_backend", meta.get("linear_kernel")),
```
[the remark this answers is not in the document]
Comment (An-mo-9) [reply benchmark/lingbot_v2/compare_thor_latency.py]: attention_backend现在在None和空的时候会报错，已修改

File: benchmark/lingbot_v2/hardware_probe.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    best_seconds = float("inf")
+    start = torch.cuda.Event(enable_timing=True)
+    end = torch.cuda.Event(enable_timing=True)
+    for _ in range(iters):
+        start.record()
+        destination.copy_(source)
+        end.record()
+        torch.cuda.synchronize(dev)
+        best_seconds = min(best_seconds, start.elapsed_time(end) / 1e3)
+    return moved / best_seconds / 1e12
```
[the remark this answers is not in the document]
Comment (An-mo-9) [reply benchmark/lingbot_v2/hardware_probe.py]: 添加了判断大于零的条件，已修改

File: phyai/src/phyai/models/lingbot_v2/scheduler_ws1_lingbotv2.py [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+        suffix_slots = suffix_slots.view(batch_size, self.suffix_len)
+        state_write = suffix_slots[:, 0]
+        action_write = suffix_slots[:, 1:].flatten()
+
+        state_indptr = torch.zeros(
+            batch_size + 1, dtype=torch.int32, device=self.device
+        )
+        state_indptr[1:] = torch.cumsum(visible_lens + 1, dim=0)
+        action_indptr = torch.zeros_like(state_indptr)
+        action_indptr[1:] = torch.cumsum(visible_lens + self.suffix_len, dim=0)
```
[the remark this answers is not in the document]
Comment (An-mo-9) [reply phyai/src/phyai/models/lingbot_v2/scheduler_ws1_lingbotv2.py]: cfg.suffix_len == cfg.chunk_size + 1是恒定的，并非两个独立的值

File: benchmark/lingbot_v2/compare_moe_thor.py [hunk 6]
```diff
@@ -0,0 +1,605 @@
...
+        raise TypeError(
+            "checkpoint routed_scaling_factor must be numeric, "
+            f"got {routed_scaling_factor!r}"
+        )
+    routed_scaling_factor = float(routed_scaling_factor)
+    if routed_scaling_factor <= 0:
+        raise ValueError(
+            "checkpoint routed_scaling_factor must be positive, "
+            f"got {routed_scaling_factor!r}"
+        )
```
[the remark this answers is not in the document]
Comment (An-mo-9) [reply benchmark/lingbot_v2/compare_moe_thor.py]: 已解决

File: phyai/src/phyai/models/lingbot_v2/main_lingbotv2.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        # The released FP32 router uses PyTorch matmul precision "high". Keep
+        # this process-wide policy at the engine boundary and restore it when
+        # this entry is closed or setup fails.
+        self._previous_matmul_precision = torch.get_float32_matmul_precision()
+        torch.set_float32_matmul_precision("high")
+        try:
+            self._setup(args)
+        except BaseException:
+            self._restore_matmul_precision()
+            raise
```
[the remark this answers is not in the document]
Comment (An-mo-9) [reply phyai/src/phyai/models/lingbot_v2/main_lingbotv2.py]: 已解决

File: phyai-utils-tools/tests/test_lingbot_v2_processor.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        step
+        for step in loaded.build_preprocessor().steps
+        if isinstance(step, NormalizerStep)
+    )
+    assert set(rebuilt_normalizer.state_dict()) == {
+        "observation.state.q01",
+        "observation.state.q99",
+        "action.q01",
+        "action.q99",
+    }
```
[the remark this answers is not in the document]
Comment (An-mo-9) [reply phyai-utils-tools/tests/test_lingbot_v2_processor.py]: 已解决
